core/src/main/java/jenkins/link/ConsoleURLProvider.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2023, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.link;
+
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.Util;
+import hudson.model.Run;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.Stapler;
+
+/**
+ * Extension point for providing console urls
+ * @since TODO
+ */
+public abstract class ConsoleURLProvider implements ExtensionPoint {
+
+    /**
+     * Provide the console url of the highest registered ordinal implementation.
+     * Defaults to {@link DefaultConsoleURLProviderImpl} otherwise.
+     */
+    public String getConsoleURL(Run<?, ?> run) {
+        return get().getConsoleURL(run);
+    }
+
+    /**
+     * Retrieve all implementations of ConsoleURLProvider.
+     */
+    public static ExtensionList<ConsoleURLProvider> all() {
+        return ExtensionList.lookup(ConsoleURLProvider.class);
+    }
+
+    /**
+     * Retrieve the highest registered ordinal implementation.
+     * @since TODO
+     */
+    public static ConsoleURLProvider get() {
+        return all().stream().findFirst().orElseThrow(() -> new RuntimeException("The DefaultConsoleURLProviderImpl should be instanciated"));
+    }
+
+    @Extension(ordinal = -100)
+    @Restricted(NoExternalUse.class)
+    public static class DefaultConsoleURLProviderImpl extends ConsoleURLProvider {
+
+        @Override
+        public String getConsoleURL(Run<?, ?> run) {
+            return Stapler.getCurrentRequest().getContextPath() + "/" + Util.encode(run.getUrl()) + "console";
```

Comment:
I think it would be nice to put the context path logic somewhere in the main class, perhaps a different static method that is the main API entrypoint, along with validation logic to make sure implementations return URLs with the expected structure (e.g. not absolute, leading/trailing slashes, whatever).

We could do something like `Functions.getActionUrl` and allow different types of URLs, but IDK if we want to do that or be more restrictive:
  https://github.com/jenkinsci/jenkins/blob/eb96b8e07bcf2da498411ede238d24a18ec075c3/core/src/main/java/hudson/Functions.java#L1888-L1904